### ISSUE TITLE
manage state between Alpheios and Arethusa

### ISSUE THREAD

**Issue by balmas** [alpheios-project/alpheios-core#221]
This issue follows up on the code review comments at 

_Originally posted by @balmas in https://github.com/alpheios-project/alpheios-core/pull/219_

**Comment by balmas** [comment]
Alpheios requirements:

1. Alpheios must display the tree icon in the toolbar and popup if the current word has treebank data
2. Alpheios must NOT display the tree icon in the toolbar and popup if the current word does not have treebank data
3. Alpheios MAY display the tree icon in the toolbar before the first lookup if the currently loaded page contains treebank references

Arethusa limitations:

1. Arethusa must be fully loaded with a document before it can respond to the getMorph API request
2. Arethusa must have its state set to the current sentence before it can respond to the getMorph api request for a word in that sentence

**Comment by balmas** [comment]
State that needs to be managed:

1. whether the page has any treebank documents
2. the list of treebank documents referenced by the page (currently only 1 supported but we want to support more per #148)
3. the treebank sentence/word reference for the current word
4. the treebank sentence/word reference for the previous word

**Comment by balmas** [comment]
![tree_workflows](https://user-images.githubusercontent.com/644870/78048068-3064b500-7347-11ea-993b-e6d085d1d5ff.png)
![load_arethusa_document](https://user-images.githubusercontent.com/644870/78048075-335fa580-7347-11ea-9c47-407d664063f0.png)

**Comment by balmas** [comment]
@kirlat when you have a minute can you take a look at the above and let me know if it makes sense and if you see any think that raises concerns?

**Comment by kirlat** [comment]
This is a very nice and detailed diagrams, thanks! 

The only question I have is about the use of `show_tree_for_page` variable. In my understanding, the only use for that is to decide whether to show a treebank icon during the initial page load, is that so? Because after the page is loaded and lookups have started it's the last word being looked up determines whether the treebank icon will appear (i.e. whether the word being looked up has any reference or not)? Of course, if the page has no treebank data at all then word lookups do not matter: we will not show the treebank icon no matter what.

Is that correct?

**Comment by balmas** [comment]
>The only question I have is about the use of show_tree_for_page variable. In my understanding, the only use for that is to decide whether to show a treebank icon during the initial page load, is that so? Because after the page is loaded and lookups have started it's the last word being looked up determines whether the treebank icon will appear (i.e. whether the word being looked up has any reference or not)? Of course, if the page has no treebank data at all then word lookups do not matter: we will not show the treebank icon no matter what.
Is that correct?

This is the point I'm wavering on.  In the above requirements I have stated:

>Alpheios MAY display the tree icon in the toolbar before the first lookup if the currently loaded page contains treebank references

But I think perhaps it's simplest for now to just say that we will only show the tree icon if the user has selected a word that has a tree, i.e. if show_tree_for_word is true and ignore show_tree_for_page for now.  In the future if we want to expand upon the controls we offer the user for finding and browsing trees associated with a document, then this decision might change.

**Comment by balmas** [comment]
Updated workflow diagram: 

![tree_workflows (1)](https://user-images.githubusercontent.com/644870/78993176-1d9e6d00-7b0b-11ea-9a5e-dffd293b13db.png)

**Comment by balmas** [comment]
The above workflow diagram reflects the availability of the following changes in the Arethusa api:

1) changes to gotoSentence which adds support for an optional list of wordIds to be passed as a parameter, which results in the supplied list of words being highlighted as selected in the tree view.

2) a new api command, findWord, which takes a word and prefix and suffix strings and returns a list of the matching wordids from the tree data.

These changes are necessary to fix #234 and #148

**Comment by balmas** [comment]
Note that  implementing this workflow means that we should be mainly using the Arethusa API via the adapter to navigate the tree display, and only updating the URL of the iFrame if the tree document to be loaded has changed.

**Comment by balmas** [comment]
@kirlat changes to Arethusa and the Treebank Template for this workflow have been deployed to the environment at https://alpheios-project.github.io/treebank-template. I'd like you to take a stab at implementing the changes on the Alpheios side.

**Comment by kirlat** [comment]
Should be solved with #257.

**Comment by balmas** [comment]
testing note: there are times when the treebankSrc is getting set to null within a page that has multiple treebank references  causing the iframe src to unload and reload. Need to investigate how this occurs.

**Comment by balmas** [comment]
Ah, found the bug. We are currently resetting the treebank url everytime the sentence changes.  It should only be reset if the *doc* changes.

E.g. in this test in LexisModule.lexicalQuery:

```
f (store.state.lexis.treebankSrc !== this._treebankDataItem.fullUrl) {
```
Will currently be true whenever the sentence changes because TreebankDataItem.fullUrl contains the sentence.

**Comment by balmas** [comment]
another thing I've noticed, which may or may not be related to the url being reset, is that sometimes the diagram icon flashes available and then unavailable for a lookup.

**Comment by kirlat** [comment]
> We are currently resetting the treebank url everytime the sentence changes. It should only be reset if the _doc_ changes.

I was under impression that we need change the URL and to reset the treebank every time any part of URL changes. And because a sentence ID is part of a URL, we need to do that when sentence changes too.

If it is not so, do we need to:

If document changes:
* Set a URL of an iframe to null 
* Set a URL to a new one
* Run a `refreshView` request until successful

If sentence changes:
* Set URL of an iframe to a new one.

Is that correct?

**Comment by kirlat** [comment]
> sometimes the diagram icon flashes available and then unavailable for a lookup.

That's because when we reset URL, we set it to `null` initially and only during a next tick we assign it a new value. We also wait for a refresh request to finish (it might take seconds in my tests) before we finally can declare a treebank to be available. In between the treebank is not available; that's why the icon is flashing.

Even though it is technically correct, it may not look nice from a user perspective. Maybe we could keep status as "available" during refreshes and set it to "unavailable" only when refresh fails? We have to only hope that nobody will click the treebank icon in between. Can we do anything better than that? What do you think?

**Comment by balmas** [comment]
> > We are currently resetting the treebank url everytime the sentence changes. It should only be reset if the _doc_ changes.
> 
> I was under impression that we need change the URL and to reset the treebank every time any part of URL changes. And because a sentence ID is part of a URL, we need to do that when sentence changes too.
> 
> If it is not so, do we need to:
> 
> If document changes:
> 
> * Set a URL of an iframe to null
> * Set a URL to a new one
> * Run a `refreshView` request until successful
> 
> If sentence changes:
> 
> * Set URL of an iframe to a new one.
> 
> Is that correct?

We only need to set the sentence in the URL the very first time a document is loaded. From then on, until the document changes, we use the API to navigate the sentences.  So we should only reset the URL of the iFrame when the document changes, not the sentence. The Sentence is part of the Url because it is needed for the initial load of the document, due to the way the Arethusa application is coded.

**Comment by balmas** [comment]
> Even though it is technically correct, it may not look nice from a user perspective. Maybe we could keep status as "available" during refreshes and set it to "unavailable" only when refresh fails? 

Yes let's try that.

**Comment by kirlat** [comment]
Should be fixed by #275

**Comment by balmas** [comment]
Alpheios Components 3.3.0-qa.20200501441

test in texts-test.alpheios.net with any treebanked text.

General regression testing of treebank functionality is required by this and other related treebank changes.

**Comment by balmas** [comment]
sigh. Something in the latest build is creating a problem. D3 in the Arethusa code is throwing an error:

`Error: <g> attribute transform: Expected number, "…e(15,15 ) scale(Infinity)".
Bo.transform @ d3.min.js:4`

I suspect it might have to do with changes to the way we set the height and width of the panel...

**Comment by balmas** [comment]
d3 IS throwing that error and it may be related to the changes I made for #243 but the main problem actually appears to be unrelated to the treebank functionality -- #295

**Comment by balmas** [comment]
the d3 error will be addressed by alpheios-project/arethusa#807. It doesn't prevent proper functioning of the treebank lookups in alpheios, it just clutters the console with an unnecessary error.

I think we can finally start testing the treebank functionality with Alpheios Components 3.3.0-qa.20200504341

**Comment by balmas** [comment]
hmm. I think this still isn't quite right. Our test on if a treebank is loaded isn't picking up changes in the document so a page with multiple documents doesn't work.

I think we need to test the document url not the provider url here.

**Comment by balmas** [comment]
that fix is in Alpheios Components 3.3.0-qa.20200514430

**Comment by monzug** [comment]
I did a bunch of lookup. except for the problem that after x times we do not have the tree icon and info, all looks good.
